Use `SET CONSTRAINTS` for `disable_referential_integrity` without superuser privileges

mtsmfm opened the pull request:
ref: 72c1557

senny commented:
Context #5523

koic commented on activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:
```diff
@@ -8,36 +8,32 @@ def supports_disable_referential_integrity? # :nodoc:
         def disable_referential_integrity # :nodoc:
           if supports_disable_referential_integrity?
-            original_exception = nil
+            tables_costraints = execute(<<-SQL).values
```
typo? `tables_constraints`

mtsmfm replied:
woops

mtsmfm replied:
fixed and force pushed :wink:

mtsmfm commented:
Sorry, I'll fix tests.

unchris commented on activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:
```diff
@@ -8,36 +8,32 @@ def supports_disable_referential_integrity? # :nodoc:
-            end
+            execute(
+              tables_constraints.collect { |table, constraint|
+                "ALTER TABLE #{quote_table_name(table)} ALTER CONSTRAINT #{constraint} DEFERRABLE"
```
Hey @mtsmfm we're trying this patch out today and discovered this particular syntax `ALTER TABLE table_name ALTER CONSTRAINT` was only added in PG 9.4. I don't know enough about this module but my guess is you'll probably want to have a fallback for people using older versions since 9.4 is pretty recent.

mtsmfm replied:
Thank you your feedback :star: 

I noticed that the syntax requires newest PG.
I'm investigating the way to alternate constraints on old PG version.

Probably we must check the PG version whether it can use `ALTER CONSTRAINT`.

mtsmfm committed:
Use `SET CONSTRAINTS` for `disable_referential_integrity` without superuser privileges

ref: 72c1557254

- We must use `authors` fixture with `author_addresses` because of its foreign key constraint.
- Tests require PostgreSQL >= 9.4.2 because it had a bug about `ALTER CONSTRAINTS` and fixed in 9.4.2.

mtsmfm commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -310,6 +310,10 @@ def supports_materialized_views?
       end
 
+      def supports_alter_constraint?
+        postgresql_version >= 90400
```
Added fallback

mtsmfm commented on .travis.yml:
```diff
@@ -50,3 +50,6 @@ services:
   postgresql: "9.4"
+  apt:
+    packages:
+    - postgresql-9.4
```
It seems that we need this package to use `alter constraint`

kamipo replied:
Is this needed even after #22993?

mtsmfm replied:
I think yes, travis will use 9.4.0.
But I'll investigate just in case.

mtsmfm replied:
I confirmed that travis uses 9.4.0.

https://github.com/mtsmfm/railstest/pull/5/commits/f23119f4b9499ec30565981198110124c8a2e1d8

If we have this hack:

https://travis-ci.org/mtsmfm/railstest/jobs/176084769/config

```
"addons": {
    "postgresql": "9.4",
    "apt": {
      "packages": [
        "postgresql-9.4"
      ]
    }
  },
```

postgres will >= 9.4.2 and we can use alter constraint:

https://travis-ci.org/mtsmfm/railstest/jobs/176084769#L564

> run PostgreSQLReferentialIntegrityWithAlterConstraintTest

But if we don't have:

https://travis-ci.org/mtsmfm/railstest/jobs/176084770/config

```
"addons": {
    "postgresql": "9.4"
  },
```

https://travis-ci.org/mtsmfm/railstest/jobs/176084770

Enabled fall back:

https://travis-ci.org/mtsmfm/railstest/jobs/176084770#L529

mtsmfm commented:
@sgrif @senny I fixed tests.

Could you review this PR?

amatsuda commented:
This looks good. Can you take a look? @sgrif @senny

andreynering commented on activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:
```diff
@@ -6,41 +6,81 @@ def supports_disable_referential_integrity? # :nodoc:
+                FROM information_schema.table_constraints
+                WHERE constraint_type = 'FOREIGN KEY'
+                AND is_deferrable = 'NO'
+                AND table_name IN (#{tables.collect { |name| quote(name) }.join(",")})
```
I think this last condition is not necessary.

mtsmfm replied:
I think so!

kamipo commented on activerecord/test/cases/associations/callbacks_test.rb:
```diff
@@ -7,7 +7,7 @@
 
 class AssociationCallbacksTest < ActiveRecord::TestCase
-  fixtures :posts, :authors, :projects, :developers
+  fixtures :posts, :authors, :author_addresses, :projects, :developers
```
I have a question. Adding `:author_addresses` fixture is related on this PR?

mtsmfm replied:
Yes.

We must use `authors` fixture with `author_addresses` because of its foreign key constraint.

kamipo replied:
I see, I realized that the context #5523. Thank you for your explanation!

kamipo commented on activerecord/test/cases/adapter_test.rb:
```diff
@@ -296,5 +279,22 @@ def setup
       end
     end
+
+    def test_disable_referential_integrity
```
Why move the test to the bottom?

mtsmfm replied:
Because we can't use `disable_referential_integrity_with_alter_constraint` within transaction.

```sql
-- setup
DROP TABLE "fk_test_has_fk";
CREATE TABLE "fk_test_has_fk" ("fk_id" integer NOT NULL);
DROP TABLE "fk_test_has_pk";
CREATE TABLE "fk_test_has_pk" ("pk_id" serial primary key);
ALTER TABLE "fk_test_has_fk" ADD CONSTRAINT "fk_name" FOREIGN KEY ("fk_id") REFERENCES "fk_test_has_pk" ("pk_id");

-- test
BEGIN;

ALTER TABLE fk_test_has_fk ALTER CONSTRAINT fk_name DEFERRABLE;

SET CONSTRAINTS ALL DEFERRED;

INSERT INTO fk_test_has_fk VALUES(1);
INSERT INTO fk_test_has_pk VALUES(1);

ALTER TABLE fk_test_has_fk ALTER CONSTRAINT fk_name NOT DEFERRABLE;
-- ERROR:  cannot ALTER TABLE "fk_test_has_fk" because it has pending trigger events
```

So we must put this test within `AdapterTestWithoutTransaction`

kamipo commented on activerecord/test/cases/adapters/postgresql/referential_integrity_test.rb:
```diff
@@ -1,7 +1,45 @@
-class PostgreSQLReferentialIntegrityTest < ActiveRecord::PostgreSQLTestCase
+if ActiveRecord::Base.connection.respond_to?(:supports_alter_constraint?) &&
+    ActiveRecord::Base.connection.supports_alter_constraint?
+class PostgreSQLReferentialIntegrityWithTest < ActiveRecord::PostgreSQLTestCase
```
`PostgreSQLReferentialIntegrityWithAlterConstraintTest`?

mtsmfm replied:
Great catch, thank you!

mtsmfm commented on activerecord/test/cases/adapters/postgresql/referential_integrity_test.rb:
```diff
@@ -1,7 +1,45 @@
+  end
+end
+else
+class PostgreSQLReferentialIntegrityWithDisableTriggerTest < ActiveRecord::PostgreSQLTestCase
```
Should we remove this test or add postgres < 9.4 to test matrix?
This test will not run on travis.

mtsmfm commented:
@sgrif rebased!
Could you review this patch?

kamipo commented on activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:
```diff
@@ -6,8 +6,50 @@ def supports_disable_referential_integrity? # :nodoc:
+
+          def disable_referential_integrity_with_alter_constraint
+            tables_constraints = execute(<<-SQL).values
+            SELECT table_name, constraint_name
```
missing indentation.

mtsmfm replied:
Wow, sorry I'll fix :bow:

kamipo commented on activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:
```diff
@@ -22,14 +64,14 @@ def disable_referential_integrity # :nodoc:
 
-              WARNING
+            WARNING
               raise e
```
Changing indentation is not needed.

parhs commented:
Any news? This is a good fix

mtsmfm commented:
I'm investigating failing tests :bowing_man:
(It may have no relation this PR)

mtsmfm commented:
Test passed!

-------------

> PG::TRDeadlockDetected: ERROR: deadlock detected

https://travis-ci.org/mtsmfm/railstest/jobs/179139936

It seems not to be related.

https://github.com/rails/rails/pull/27140#issuecomment-262157635

parhs commented:
Cool. So this will be available in rails 5?

parhs commented:
Also why alter table is done outside transaction?

mtsmfm commented:
> why alter table is done outside transaction?

Because it seems that we must commit ALTER TABLE to enable `SET CONSTRAINTS ALL DEFERRED;`

https://github.com/rails/rails/pull/21233#r58213937

yahonda commented:
I had accidentally closed pull request #25328 by deleting the corresponding branch. I was thinking to re-open this pull request and found #21233 addresses the original issue reported in #25328 since it changes Fixtures to load `:authors` and `:author_addresses` at the same transaction.

To address #25328 both `:authors` and `:author_addresses` fixtures need to be loaded since there is a foreign key created from `:authors` to `:author_addresses`.

https://github.com/rails/rails/blob/master/activerecord/test/schema/schema.rb#L55-L66

Also when enabling foreign key at the end of fixture loading, Oracle database also checks the existing data before enabling foreign key. MySQL does not. Please note that I do not mean to say which one is good or bad. I meant to say that users (in this case ActiveRecord fixture developers) have responsibility to keep the data consistency while disabling foreign key, then loading both `:authors` and `:author_addresses` fixtures make a lot of sense.

It would be appreciated this pull request reviewed and merged to master branch.

Thanks.

mtsmfm commented:
Thanks!

rafaelfranca commented:
I'm reverting this commit because it broke the PostgreSQL isolated tests https://travis-ci.org/rails/rails/builds/188721346.

Could you take a look and fix those tests. You can run them locally with `rake `test:isolated:postgresql`?

mtsmfm commented:
@rafaelfranca 

Sorry about that 🙇 

I ran whole tests on travis in my repo:

https://github.com/mtsmfm/railstest/pull/5
https://travis-ci.org/mtsmfm/railstest/builds/180910862

But I probably got something wrong 😢 
I'll investigate this failing tests 💦 

----

BTW, travis didn't run tests for this PR?
It seems that only Code Climate worked for this commit:

![image](https://cloud.githubusercontent.com/assets/1796864/21632046/72b31232-d285-11e6-81dc-1049cee9a4ec.png)

yahonda commented:
I think I am able to reproduce [ActiveRecord unit test errors and failures](https://travis-ci.org/rails/rails/jobs/188721371). 

```
$ ruby -v
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-linux]
```

```
$ psql -d activerecord_unittest
psql (9.6.1)
Type "help" for help.

activerecord_unittest=# select version();
                                                 version                                                 
---------------------------------------------------------------------------------------------------------
 PostgreSQL 9.6.1 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 6.2.1 20160916 (Red Hat 6.2.1-2), 64-bit
(1 row)

activerecord_unittest=# 
```



* Branch to reproduce https://github.com/yahonda/rails/tree/chery_pick_21233

```ruby
$ git checkout chery_pick_21233
$ bundle
$ ARCONN=postgresql bundle exec ruby -Itest test/cases/associations/left_outer_join_association_test.rb 
Using postgresql
Run options: --seed 35633

# Running:

WARNING:  there is no transaction in progress
E....EF..E..F

Finished in 0.238527s, 54.5011 runs/s, 46.1163 assertions/s.

  1) Error:
LeftOuterJoinAssociationTest#test_left_outer_joins_return_has_value_for_every_comment:
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  insert or update on table "authors" violates foreign key constraint "fk_rails_94423a17a3"
DETAIL:  Key (author_address_id)=(1) is not present in table "author_addresses".
: COMMIT
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:97:in `async_exec'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:97:in `block (2 levels) in execute'
    /home/yahonda/git/rails/activesupport/lib/active_support/dependencies/interlock.rb:46:in `block in permit_concurrent_loads'
    /home/yahonda/git/rails/activesupport/lib/active_support/concurrency/share_lock.rb:185:in `yield_shares'
    /home/yahonda/git/rails/activesupport/lib/active_support/dependencies/interlock.rb:45:in `permit_concurrent_loads'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:96:in `block in execute'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:594:in `block in log'
    /home/yahonda/git/rails/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:587:in `log'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:95:in `execute'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:165:in `commit_db_transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:140:in `commit'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:174:in `commit_transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:199:in `within_new_transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:225:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:38:in `disable_referential_integrity_with_alter_constraint'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/postgresql/referential_integrity.rb:12:in `disable_referential_integrity'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:523:in `create_fixtures'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:1025:in `load_fixtures'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:965:in `setup_fixtures'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:851:in `before_setup'


  2) Error:
LeftOuterJoinAssociationTest#test_construct_finder_sql_applies_aliases_tables_on_association_conditions:
ActiveRecord::RecordNotFound: Couldn't find Author with 'id'=1
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:350:in `raise_record_not_found_exception!'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:467:in `find_one'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:447:in `find_with_ids'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:66:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/querying.rb:3:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/core.rb:171:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:837:in `block in find'
    /home/yahonda/git/rails/activerecord/lib/active_record/scoping/default.rb:35:in `block in unscoped'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation.rb:341:in `scoping'
    /home/yahonda/git/rails/activerecord/lib/active_record/scoping/default.rb:35:in `unscoped'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:836:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:920:in `block (4 levels) in setup_fixture_accessors'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:915:in `map'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:915:in `block (3 levels) in setup_fixture_accessors'
    test/cases/associations/left_outer_join_association_test.rb:14:in `test_construct_finder_sql_applies_aliases_tables_on_association_conditions'


  3) Failure:
LeftOuterJoinAssociationTest#test_does_not_override_select [test/cases/associations/left_outer_join_association_test.rb:77]:
Failed assertion, no message given.


  4) Error:
LeftOuterJoinAssociationTest#test_find_with_sti_join:
ActiveRecord::RecordNotFound: Couldn't find Post with 'id'=4
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:350:in `raise_record_not_found_exception!'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:467:in `find_one'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:447:in `find_with_ids'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation/finder_methods.rb:66:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/querying.rb:3:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/core.rb:171:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:837:in `block in find'
    /home/yahonda/git/rails/activerecord/lib/active_record/scoping/default.rb:35:in `block in unscoped'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation.rb:341:in `scoping'
    /home/yahonda/git/rails/activerecord/lib/active_record/scoping/default.rb:35:in `unscoped'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:836:in `find'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:920:in `block (4 levels) in setup_fixture_accessors'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:915:in `map'
    /home/yahonda/git/rails/activerecord/lib/active_record/fixtures.rb:915:in `block (3 levels) in setup_fixture_accessors'
    test/cases/associations/left_outer_join_association_test.rb:67:in `test_find_with_sti_join'


  5) Failure:
LeftOuterJoinAssociationTest#test_left_outer_joins_count_is_same_as_size_of_loaded_results [test/cases/associations/left_outer_join_association_test.rb:28]:
Expected: 17
  Actual: 0

13 runs, 11 assertions, 2 failures, 3 errors, 0 skips
$
``` 

Possible fix is adding `:author_addresses` to [left_outer_join_association_test.rb](https://github.com/yahonda/rails/commit/bba722f5706c85ea69a35d90d71721707eb0d47b) . This branch is also pushed to https://github.com/yahonda/rails/tree/chery_pick_21233_and_add_fix

Failures at railties will not reproduce since https://github.com/rails/rails/commit/35404ae046b49705c412920917f92f2adbec5f59 has been removed.

mtsmfm commented:
@yahonda Thank you for your information!

I reproduced the failure on machine.

> Possible fix is adding :author_addresses to left_outer_join_association_test.rb .

I think so.
I'll cherry-pick your commit and re-create PR.
I forgot to add `:author_addresses` on left_outer_join_association_test.rb 🙇 

I confirmed there aren't any other missing by following snippet:

```diff
--- a/activerecord/lib/active_record/fixtures.rb
+++ b/activerecord/lib/active_record/fixtures.rb
@@ -889,6 +889,9 @@ def set_fixture_class(class_names = {})
       end

       def fixtures(*fixture_set_names)
+        if fixture_set_names.include?(:authors) && fixture_set_names.exclude?(:author_addresses)
+          raise caller.first
+        end
         if fixture_set_names.first == :all
           fixture_set_names = Dir["#{fixture_path}/{**,*}/*.{yml}"].uniq
           fixture_set_names.map! { |f| f[(fixture_path.to_s.size + 1)..-5] }
```

-------------------

> I ran whole tests on travis in my repo:
>
> https://github.com/mtsmfm/railstest/pull/5
> https://travis-ci.org/mtsmfm/railstest/builds/180910862

I got the cause.
It seems that the failure depends on test order.
We can pass the test if we run tests which has :authors and :author_addresses before left_outer_join_association_test because of its cache (https://github.com/rails/rails/blob/b292b76c2dd0f04fb090d49b90716a0e6037b41a/activerecord/lib/active_record/fixtures.rb#L518-L520)

```
ARCONN=postgresql bundle exec ruby -I"lib:test" -I"/tmp/vendor/bundle/gems/rake
-12.0.0/lib" "/tmp/vendor/bundle/gems/rake-12.0.0/lib/rake/rake_test_loader.rb" "test/cases/associations_test.rb" "t
est/cases/associations/left_outer_join_association_test.rb"
Using postgresql
Run options: --seed 49015

# Running:

......................................................

Finished in 0.870522s, 62.0318 runs/s, 106.8325 assertions/s.

54 runs, 93 assertions, 0 failures, 0 errors, 0 skips
```

yahonda commented:
Great.  

To isolate each unit test file, you can execute ActiveRecord unit test like this: 
```
$ bundle exec rake test:isolated
```

This syntax is used at Rails CI at travis, then it may be safer to run unit test at your side before submitting pull requests.

mtsmfm commented:
> This syntax is used at Rails CI at travis

Wow, I didn't know the difference between PR's build and master.

https://github.com/rails/rails/blob/b292b76c2dd0f04fb090d49b90716a0e6037b41a/ci/travis.rb#L148

Thank you for useful information!
I'll run `rake test:isolated` instead of `rake test:postgresql` before submitting PR and remove the line on my fork repo!
